File: nextjs-app/src/app/types/chatCompletions.ts
export type FunctionCall = {
  name: string,
  arguments: string
}
export type ToolCall = {
  id: string,
  type: "function",
  function: FunctionCall 
}
export type ToolResponseMessage = {
  role: 'tool',
  content: string,
  name: string, // function name
  tool_call_id: string, // given by OR
  _createdAt: string,
}
export type NormalMessage = {
  role: string,
  content: string,
  tool_calls: ToolCall[] | null,
  _createdAt: string
}
export type Message = NormalMessage | ToolResponseMessage

export type ChatCompletionChoice = {
  message: Message
}
export type ChatCompletion = {
  id: string | null,
  choices: ChatCompletionChoice[] | null
}

export type ChatCompletionTool = {
  type: 'function';
  function: {
    name: string;
    description: string;
    parameters: {
      type: 'object';
      properties: {
        [key: string]: {
          type: string;
          description: string;
        };
      };
      required: string[];
    };
  };
};

// Defines the overall structure of the request body sent to the AI service
export type ChatCompletionRequestBody = {
  model: string;
  tools: ChatCompletionTool[];
  messages: Message[];
};

export type UserTextSubmissionEvent = {
  type: 'UserTextSubmissionEvent';
  text: string;
  timestamp: number; 
}
export type UserTextSubmissionReceiptEvent = {
  type: 'UserTextSubmissionReceiptEvent';
  text: string;
  timestamp: number; 
}
export type RequestAssistantMessageEvent = {
  type: 'RequestAssistantMessageEvent';
  text: 'ready',
  timestamp: number;
}
export type AssistantMessageGenerationStartedEvent = {
  type: 'AssistantMessageGenerationStartedEvent';
  text: 'generating',
  timestamp: number;
}
export type AssistantMessageEvent = {
  type: 'AssistantMessageEvent';
  completion: ChatCompletion
  timestamp: number;
}
export type Event = AssistantMessageEvent | AssistantMessageGenerationStartedEvent | UserTextSubmissionEvent | UserTextSubmissionReceiptEvent | RequestAssistantMessageEvent; 
export type EventContainer = {
  event: Event
  prevEventHash: string | null,
  currentEventHash: string
}

File: parfit/shared/src/chat_completions.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FunctionCall {
    pub name: String,
    pub arguments: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolCall {
    pub id: String,
    #[serde(rename = "type")]
    pub tool_type: String, // "function"
    pub function: FunctionCall,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolResponseMessage {
    pub role: String, // "tool"
    pub content: String,
    pub name: String,
    pub tool_call_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AssistantMessage {
    pub role: String,
    pub content: String,
    pub tool_calls: Option<Vec<ToolCall>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserMessage {
  pub role: String,
  pub content: String
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SystemMessage {
  pub role: String,
  pub content: String
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
// most specific versions first
pub enum ChatMessage {
  Assistant(AssistantMessage),
  Tool(ToolResponseMessage),
  User(UserMessage),
  System(SystemMessage),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatCompletionChoice {
    pub message: ChatMessage,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatCompletion {
    pub id: Option<String>,
    pub choices: Option<Vec<ChatCompletionChoice>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FunctionSchema {
    pub name: String,
    pub description: String,
    pub parameters: FunctionParameters,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FunctionParameters {
    #[serde(rename = "type")]
    pub param_type: String, // "object"
    pub properties: HashMap<String, PropertyDefinition>,
    pub required: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PropertyDefinition {
    #[serde(rename = "type")]
    pub prop_type: String,
    pub description: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatCompletionTool {
    #[serde(rename = "type")]
    pub tool_type: String, // "function"
    pub function: FunctionSchema,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatCompletionRequestBody {
    pub model: String,
    pub tools: Option<Vec<ChatCompletionTool>>,
    pub messages: Vec<ChatMessage>,
}
